test/unit/org/apache/cassandra/index/IndexStatusManagerTest.java

```diff
@@ -0,0 +1,393 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import org.apache.cassandra.exceptions.ReadFailureException;
+import org.apache.cassandra.gms.VersionedValue;
+import org.apache.cassandra.locator.*;
+import org.apache.cassandra.utils.JsonUtils;
+import org.junit.Test;
+
+import org.apache.cassandra.db.ConsistencyLevel;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.schema.IndexMetadata;
+import org.mockito.Mockito;
+
+import static org.apache.cassandra.locator.ReplicaUtils.full;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.Assert.assertArrayEquals;
+
+public class IndexStatusManagerTest
+{
+    static class Testcase
+    {
+        String keyspace;
+        int minRequiredNodes;
+        AbstractReplicationStrategy replicationStrategy;
+        Map<InetAddressAndPort, Map<String, Index.Status>> indexStatus;
+        EndpointsForRange expected;
+
+        Testcase(Builder builder)
+        {
+            keyspace = builder.keyspace;
+            minRequiredNodes = builder.minRequiredNodes;
+            replicationStrategy = builder.replicationStrategy;
+            indexStatus = builder.indexStatus;
+            expected = builder.expected;
+        }
+
+        static class Builder
+        {
+            String keyspace;
+            int minRequiredNodes;
+            AbstractReplicationStrategy replicationStrategy;
+            Map<InetAddressAndPort, Map<String, Index.Status>> indexStatus;
+            EndpointsForRange expected;
+
+            Builder keyspace(String kp)
+            {
+                keyspace = kp;
+                return this;
+            }
+
+            Builder minRequiredNodes(int required)
+            {
+                minRequiredNodes = required;
+                return this;
+            }
+
+            Builder replicationStrategy(AbstractReplicationStrategy strategy)
+            {
+                replicationStrategy = strategy;
+                return this;
+            }
+
+            Builder indexStatus(Map<InetAddressAndPort, Map<String, Index.Status>> status)
+            {
+                indexStatus = status;
+                return this;
+            }
+
+            Builder expected(EndpointsForRange endpoints)
+            {
+                expected = endpoints;
+                return this;
+            }
+
+            Testcase build()
+            {
+                return new Testcase(this);
+            }
+        }
+    }
+
+    // we have enough BUILD_SUCCEEDED endpoints: filterForQuery should prioritize them, even if the other endpoints'
+    // indexes are queryable but not built (like UNKNOWN)
+    @Test
+    public void filterForQuery_ShouldPrioritizeSuccessfulEndpoints()
+    {
+        runTest("exactly enough endpoints with BUILD_SUCCEEDED indexes",
```

Comment:
I'd make this its own top-level test, called perhaps `shouldPrioritizeSingleSuccessfulEndpoint()`